### ISSUE TITLE
Prediction object should handle missing values in response/prob

### ISSUE THREAD

**Issue by Coorsaa** [mlr-org/mlr3#118]
`predict.glm()` predicts an `NA` if data contains missing values (`na.action` default). As a result you can get
````
response
 [1] <NA> R    R    R    R    R    R    R    R    R    M    M    M   
[14] M    M    M    M    M    M    M   
Levels: M R
````
as response vector. However, when creating the PredictionClassif object, you get the following error:
````
> PredictionClassif$new(e$task, response, prob)
Error in predictionclassif_initialize(self, task, response, prob) : 
  Assertion on 'response' failed: Contains missing values (element 1). 
````
Of course, the same applies for regression task prediction objects.

**Comment by mllg** [comment]
One could argue that this should be an error. Just removing them from the score does not seem right. 

We could either use the fallback learner to provide some predictions, or require the user the set the necessary hyperparameters by raising such an exception.

@berndbischl ?

**Comment by berndbischl** [comment]
i dont even get the question, sorry :)
let me try to rephrase it precisely.

so we use glm, and the *input* contains NAs`
and we agree that glm has no smart, native miss-value-handling mechanism?
so it does not have the missing property?

so *passing in* missings violates the contract?

isnt then the consequence pretty obvious?
its an error or undefined behavior. an error is always better, and it should be as informative as possible.
in this case, hey, what about: "glm doesnt handle missing values"

now the only question is, whether michel wants to invest the computation time to check (larger) dataframes for NAs.
my personal opinion: mlr3 should check for such stuff, and i can use an option to switch that off to get a speed improvement (if time is not negligible, otherwise noone needs the option...)

should we use the fallback learner:  no? the fallback learner is for instances which are hard to foresee and hard to avoid. this is neither? because its like this:

- hey i can check myself for NAs in my data
- now i know glm doesnt work
- i can impute to repair

**Comment by mllg** [comment]
If I understood Stefan correctly, we are talking about a split into train/test where the train set does not contain any NAs (so the model fit works), but the test set has NAs which leads to predictions being NA.

Regarding checking for NA: It is still unclear how this will look like for pipelines. A PipeOp can impute, but I can never be sure if the OP really imputes *all* missing values. We can discuss this offline next week.

Either way, now that we have a fallback mechanism in place, we could use it also for NA predictions. Is there a downside?

**Comment by berndbischl** [comment]
> If I understood Stefan correctly, we are talking about a split into train/test where the train set does not contain any NAs (so the model fit works), but the test set has NAs which leads to predictions being NA.

sorry! i didnt get this @Coorsaa 
(please really a few more details to explain this, in general)

this is exactly the point we discussed in munich, in my office right?
well now my opinion changes a bit and is: 
it is still a violation of the contract, and you can produce an error, it just "sucks" if just one obs violates, but i still get nothing for the other - although I easily could. 


> Regarding checking for NA: It is still unclear how this will look like for pipelines. A PipeOp can impute, but I can never be sure if the OP really imputes _all_ missing values. We can discuss this offline next week.
> 
> Either way, now that we have a fallback mechanism in place, we could use it also for NA predictions. Is there a downside?

i dont get what you mean. can you please define this precisely?

so, here is the situation. we have 
x = 1:1000
x[500] = NA

what is f(x)? 

PS: 
> t is still unclear how this will look like for pipelines. A PipeOp can impute, but I can never be sure if the OP really imputes _all_ missing values.

It is mainly unclear, because we dont write specs docs, for stuff like this, so we disscuss the same point 3 times (at least IMHO...)
And why can't "a pipe impute _all_ missing values"?

**Comment by berndbischl** [comment]
in order to be constructive:

a) in my example above i am _guessing_ you would impute f(x[500]) with a prediction from the fallback learner? so "mix" that into the prediction vector, to make it complete? 
(instead of producing 1000 predictions from the FB learner - which would necessarily suck IMHO)

my answer: good idea. I just dislike the "special handling" in mlr3, and in mlr we had too much of this stuff. but i could live with this. my point in munich was simply to have this in pipelines to have a cleaner base-mlr3 - and make the imputation encapsulated and very transparent to the user.

if we move forward with the discussion here, i am happy to write down the idea for the PipeOp again? and then we can decide?

**Comment by berndbischl** [comment]
> Ok, fair enough. Just saying that imputing predictions using the fallback would be very simple and useful.

then do them now and we keep a mini issue open? to remove the feature _if_ it becomes clear that this is better in pipelines? i dont want to block this

**Comment by berndbischl** [comment]
or we just wait. this is not super urgent?

**Comment by mb706** [comment]
Now there exists a [`PipeOpBackupLearner`](https://github.com/mlr-org/mlr3pipelines/blob/master/R/PipeOpBackupLearner.R) which can [handle and repair missings from predictions](https://github.com/mlr-org/mlr3pipelines/blob/7efe7e0d2dd32f53fe0b9b8a16cea8edcd89682d/tests/testthat/test_pipeop_backup_learner.R#L77-L81). For this to work, the `Prediction` creator should not throw an error when `NA`s are predicted (which IMHO is not a good idea regardless).

In fact, there are learners like `classif.kknn` that seem to just skip rows with `NA`s during prediction (leading to `PredictionClassif$new` complaining that not enough rows were predicted), they should be repaired to return predictions with `NA`s.

**Comment by mllg** [comment]
The fallback-learner mechanism is now removed from mlr3, I'll look into extending `PipeOpBackupLearner` to handle all sorts of misbehaving learners.

**Comment by mllg** [comment]
This is now covered via #258. Missing predictions of the learner are imputed with a fallback learner. 

@mb706 I opted to include this into mlr3 instead of making it a separate PipeOp because I believe that this is a very frequent problem. We can discuss this in Toulouse, if you like.
